mfs/ops.go

```diff
@@ -12,60 +13,175 @@ import (
 	cid "gx/ipfs/Qmdu2AYUV7yMoVBQPxXNfe7FJcdx16kYtsx6jAPKWQYF1y/go-cid"
 )
 
+var (
+	errInvalidNodeType = errors.New("invalid argument")
+	errMvDirToFile     = errors.New("can not move directory to file path")
+	errMvParentDir     = errors.New("can not move parent directory to sub directory")
+	errInvalidDirPath  = errors.New("path end with '/' is not a directory")
+)
+
 // Mv moves the file or directory at 'src' to 'dst'
 func Mv(r *Root, src, dst string) error {
-	srcDir, srcFname := gopath.Split(src)
+	if src == "/" {
+		return errMvParentDir
+	}
+
+	src = strings.TrimRight(src, "/")
+	if strings.HasPrefix(dst, src) {
+		return errMvParentDir
+	}
+
+	srcNode, err := DirLookup(r.GetDirectory(), src)
+	if err != nil {
+		return err
+	}
 
-	var dstDirStr string
-	var filename string
-	if dst[len(dst)-1] == '/' {
-		dstDirStr = dst
-		filename = srcFname
+	if srcNode.Type() == TDir {
+		return moveDir(r, src, dst)
+	} else if srcNode.Type() == TFile {
+		if src[len(src)-1] == '/' {
```

Comment:
The trailing `/` gets trimmed above. We should probably also have a test for this.

Reply:
fixed